affinity_ocata/README.rst

```diff
@@ -0,0 +1,308 @@
+If there would be not possible to put an instance together on an
+aggregate (in case of affinity) or on different one (in case of
+anti-affinity), it will be placed in specified group anyway.
+
+Simple usage is as follows, using environment described above in
+*aggregate-affinity* feature:
+
+.. code:: shell-session
+
+   $ export OS_COMPUTE_API_VERSION=2.43
+   $ openstack aggregate create rack1
+   $ openstack aggregate create rack2
+   $ openstack aggregate add host rack1 $(openstack baremetal node list|grep node-0|awk '{print $2}')
+   $ openstack aggregate add host rack1 $(openstack baremetal node list|grep node-1|awk '{print $2}')
+   $ openstack aggregate add host rack2 $(openstack baremetal node list|grep node-2|awk '{print $2}')
+   $ openstack aggregate add host rack2 $(openstack baremetal node list|grep node-3|awk '{print $2}')
+   $ openstack server group create --policy aggregate-soft-anti-affinity group1
+   $ openstack server create \
+         --image=$(openstack image list|grep x86_64-disk| awk '{print $2}') \
+         --flavor=1 \
+         --nic net-id=$(openstack network list |grep private | awk '{print $2}') \
+         --hint group=$(openstack server group list | grep group1 | awk '{print $2}') \
+         instance1
+   $ openstack server create \
+         --image=$(openstack image list|grep x86_64-disk| awk '{print $2}') \
+         --flavor=1 \
+         --nic net-id=$(openstack network list |grep private | awk '{print $2}') \
+         --hint group=$(openstack server group list | grep group1 | awk '{print $2}') \
+         instance2
+   $ openstack server create \
+         --image=$(openstack image list|grep x86_64-disk| awk '{print $2}') \
+         --flavor=1 \
+         --nic net-id=$(openstack network list |grep private | awk '{print $2}') \
+         --hint group=$(openstack server group list | grep group1 | awk '{print $2}') \
+         instance3
+
+Unlike in ``aggregate-anti-affinity`` policy, creating ``instance3`` will
+pass, since regardless of not available aggregate with no group members, it
+will be placed in the group anyway on one of the available host within the
+group.
+
+
+Configuration
+-------------
+
+As for soft aggregate (anti) affinity there is another limitation, which comes
+with how weights works right now in Nova. Because of `this commit`_ change of
+behaviour was introduced on how scheduler selects hosts. It's concerns all of
+affinity/anti-affinity weights, not only this particular newly added for
+aggregation.
+
+That change introduce a blind selection of the host form a group of the weighed
+hosts, which are originally sorted from best fitting. For affinity weight it
+will always return full list of the hosts (since they are not a filters), which
+is ordered from best to worst hosts. There is a high chance, that ``nova.conf``
+will need to have a scheduler filter option ``host_subset_size`` set to ``1``,
+like:
+
+.. code:: ini
+
+   [filter_scheduler]
+   host_subset_size = 1
```

Comment:
Had a conversation in Slack about this - @gryf has some ideas on fixing it and is going to write some code tomorrow.

```
18:51:17         jroll | hey y'all, affinity stuff is looking good so far
18:51:23         jroll | this concerns me a bit: https://github.com/yahoo/openstack/pull/2/files#diff-6d511686793305eaaaca7c10966bd46eR112 | Group Ironic nodes to be associate with aggregates by gryf · Pull Request #2 · yahoo/openstack
18:51:59         jroll | mostly due to this (from ironic docs): ```
18:51:59               | # Size of subset of best hosts selected by scheduler.
18:51:59               | # New instances will be scheduled on a host chosen randomly from a
18:51:59               | # subset of the 999 hosts. The big value is used to avoid race
18:51:59               | # conditions, when several instances are scheduled on the same bare
18:51:59               | # metal nodes. This is not a problem when resource classes are used.
18:51:59               | host_subset_size=999
18:51:59               | ```
18:53:00          gryf | yea. there is a link in commit message to the work, where someone need that
18:54:04          gryf | this thing is working like this - if host_subset_size is set to something larger than 1, than affinity (for aggregate or for vms) weight probably select wrong host
18:55:33          gryf | because logic behind this is to select not the best host which weight will provide, but random host from the set weight provide, even though all the other host might have negative weight.
18:55:59          gryf | devstack set this option to 999
18:56:08          gryf | I think that default is 1  (edited)
18:56:16         jroll | but also, because of an (undocumented, apparently?) problem with how ironic presents nodes to the scheduler - they're always in the same order, so in an environment where nodes are recycled often, the same node will get hit often for scheduling
18:56:25         jroll | leading to some nodes wearing out faster than others
18:56:31         jroll | (sorry, got pulled away for a minute)
18:56:51          gryf | yeah, that might be an issue
18:56:52         jroll | yeah, I understand the problem, I'm not sure the best route to fix it. but wanted to bring it up and get people thinking about it
18:57:28         jroll | this is only with soft affinity, right, because it's a weigher instead of a filter?
18:57:34          gryf | yes
18:57:43         jroll | ok cool
18:58:03          gryf | aggregate filters will remove hosts from the list
18:58:20         jroll | yep, makes sense, just double checking
18:59:00          gryf | I think, that this issue might be worked around by first selecting the best matches (with random pick) and if there is no suitable host, than select from the worst matches
19:00:07          gryf | I can cook up a quick fix for that tomorrow
19:03:29          gryf | I'm just wondering, that no one ever complain about how soft-affinity/soft-anti-affinity work after that patch, since even though there are compute hosts which can host the vms for affinity, the other compute will be selected…
19:03:41         jroll | @gryf: I'd be interested to see what you're thinking there, in code
19:04:03          gryf | ok than. I'll prepare some patch tomorrow
19:04:03         jroll | yeah, I agree the best behavior is with `host_subset_size=1`
19:04:21         jroll | feel free to just paste a diff in here if it's small, I just don't quite get what you plan to change
19:04:33          gryf | :+1:
19:04:45         jroll | thanks!
19:04:50          gryf | np :D
```

Reply:
This is handled by an upstream commit in queens - https://review.openstack.org/#/c/494136/

Going to backport that internally but leave it off of this PR.